crates/astria-sequencer-relayer/src/relayer/write.rs

```diff
@@ -0,0 +1,366 @@
+    jsonrpsee::http_client::HttpClient,
+    submission::ToBlobs as _,
+};
+use futures::{
+    future::{
+        Fuse,
+        FusedFuture as _,
+    },
+    FutureExt as _,
+};
+use pin_project_lite::pin_project;
+use sequencer_client::{
+    tendermint::block::Height as SequencerHeight,
+    SequencerBlock,
+};
+use tokio::{
+    select,
+    sync::mpsc::{
+        self,
+        error::{
+            SendError,
+            TrySendError,
+        },
+        Receiver,
+        Sender,
+    },
+};
+use tokio_util::task::JoinMap;
+use tracing::{
+    debug,
+    error,
+    info,
+    instrument,
+    warn,
+    Instrument as _,
+    Span,
+};
+
+struct QueuedBlobs {
+    blobs: Vec<Blob>,
+    heights: Vec<SequencerHeight>,
+}
+
+impl QueuedBlobs {
+    fn num_blobs(&self) -> usize {
+        self.blobs.len()
+    }
+
+    fn new() -> Self {
+        Self {
+            heights: Vec::new(),
+            blobs: Vec::new(),
+        }
+    }
+
+    fn push(&mut self, mut blobs: Vec<Blob>, height: SequencerHeight) {
+        self.blobs.append(&mut blobs);
+        self.heights.push(height);
+    }
+
+    fn take(&mut self) -> TakeBlobs<'_> {
+        TakeBlobs {
+            queue: Some(self),
+        }
+    }
+}
+
+pin_project! {
+    struct TakeBlobs<'a> {
+        queue: Option<&'a mut QueuedBlobs>,
+    }
+}
+
+impl<'a> Future for TakeBlobs<'a> {
+    type Output = Option<(Vec<Blob>, Vec<SequencerHeight>)>;
+
+    fn poll(self: std::pin::Pin<&mut Self>, _: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let queue = self
+            .project()
+            .queue
+            .take()
+            .expect("this future must not be polled twice");
+        let blobs = std::mem::take(&mut queue.blobs);
+        let heights = std::mem::take(&mut queue.heights);
+        if blobs.is_empty() {
+            Poll::Ready(None)
+        } else {
+            Poll::Ready(Some((blobs, heights)))
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(super) struct BlobSubmitterHandle {
+    tx: Sender<SequencerBlock>,
+}
+
+impl BlobSubmitterHandle {
+    // allow: just forwarding the error type
+    #[allow(clippy::result_large_err)]
+    pub(super) fn try_send(
+        &self,
+        block: SequencerBlock,
+    ) -> Result<(), TrySendError<SequencerBlock>> {
+        self.tx.try_send(block)
+    }
+
+    // allow: just forwarding the error type
+    #[allow(clippy::result_large_err)]
+    pub(super) async fn send(
+        &self,
+        block: SequencerBlock,
+    ) -> Result<(), SendError<SequencerBlock>> {
+        self.tx.send(block).await
+    }
+}
+
+pub(super) struct BlobSubmitter {
+    // The client to submit blobs to Celestia.
+    client: HttpClient,
+
+    // The channel over which sequencer blocks to receive.
+    blocks: Receiver<SequencerBlock>,
+
+    // The maximum number of conversions that can be active at the same time.
+    // Submitter will refuse to accept more blocks until there are fewer active
+    // conversions.
```

Comment:
what happens to blocks that are refused? are they resent to the submitter later?

Reply:
Yes, have a look at the usage of `try_send` vs `send` in the relayer select-loop. It always `try_send`s a block first. If that returns an error it immediately schedules a future with `send(the_same_block)` and pauses the block stream.